File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.py
#1시반
def solution(record):
    answer = []
    in_out = []
    print_lis = {'Enter':"님이 들어왔습니다.",'Leave':"님이 나갔습니다."}
    dict = {}  #id : 닉네임
    for i in range(len(record)):
        temp = record[i].split(" ")
        if temp[0] == "Enter" or temp[0] == "Change":
            dict[temp[1]] = temp[2]
        in_out.append([temp[1],temp[0]])
    # print(in_out)
    for i in in_out:
        #아이디 enter
        if i[1] == 'Change':
            continue
        answer.append(dict[i[0]]+print_lis[i[1]])
    return answer

#닉네임 변경 -> 기존에 출력된 닉네임도 변경
# 나가고 이름 바꿔서 들어오면 출력된 닉넴 변경

#record: 닉네임 변경 기록
#최종 출력 : return
#Enter, Leave, Change

File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.java
import java.util.*;

class Solution {
    static class Msg {
        String id;
        String action; 
        
        public Msg(String id, String action) {
            this.id = id;
            this.action = action;
        }
    }
    public String[] solution(String[] record) {
        ArrayList<String> answer = new ArrayList<>();
        
        HashMap<String, String> nickid = new HashMap<>();

        HashMap<String, String> msg = new HashMap<>();
        msg.put("Enter", "님이 들어왔습니다.");
        msg.put("Leave", "님이 나갔습니다." );
        
        ArrayList<Msg> list = new ArrayList<>();
        for (String str : record) {
            String[] spl = str.split(" ");
            String id = spl[1];
            if (spl[0].equals("Change")) {
                String name = spl[2];
                nickid.put(id,name);
            }
            else if (spl[0].equals("Enter")) {
                String name = spl[2];
                nickid.put(id,name);
                list.add(new Msg(id, spl[0]));
            }
            else {
                list.add(new Msg(id, spl[0]));
            }
        }
        
        StringBuilder sb = new StringBuilder();        
        for (Msg node : list) {
            String id = node.id;
            String action = node.action;
            
            answer.add(nickid.get(id) + msg.get(action));
        }
        
        return answer.toArray(new String[0]);
    }
}